### ISSUE TITLE
Unable to scan library in Web-INF/ Lib

### ISSUE THREAD

**Issue by surajanuja** [CycloneDX/cyclonedx-gradle-plugin#296]
@stevespringett - Is there a way to scan the legacy libraries which are present in Web-INF/Lib but not part of build.gradle.

Currently when I use cyclonedx, it generates bom for all the libraries defined in build.gradle but if there is any library not part of build.gradle, it excludes those libraries from BOM

So I want to know if there is any way to define a path in build.grqdle so that cyclonedx will generate bom with libraries that is present in that path? Or is there any other solution that you can recommend?

**Comment by stevespringett** [comment]
I don't think that's possible today with the gradle plugin. @prabhu, does cdxgen support this?

**Comment by prabhu** [comment]
@stevespringett Yes, cdxgen supports `-t jar` mode. However, it is not possible to combine this with `java`, so needs some enhancement and testing.

https://github.com/CycloneDX/cdxgen/blob/ef453ca771c4481b771f5dcd8db9aaef44025abe/index.js#L897

**Comment by surajanuja** [comment]
> @stevespringett Yes, cdxgen supports `-t jar` mode. However, it is not possible to combine this with `java`, so needs some enhancement and testing.
> 
> https://github.com/CycloneDX/cdxgen/blob/ef453ca771c4481b771f5dcd8db9aaef44025abe/index.js#L897

@prabhu My requirement is only to scan jar libraries from build.gradle + the libraries which are present in specific path which is not part of build.gradle. 

I tried using cdxgen also but still it is also not scanning any library outside build.gradle. I dont know if I am missing any settings that I have to add in build.gradle to scan a path

**Comment by prabhu** [comment]
@tech-surajrk Could you share the cdxgen command used? Did you specify `-t jar`?

**Comment by tech-surajrk** [comment]
> @tech-surajrk Could you share the cdxgen command used? Did you specify `-t jar`?

@prabhu using '-t jar ' - I am able to generate bom but the group name for each component is coming as undefined or blank. Hence Dependency Track is not able to map any cve against the component
Is there any way to generate correct group name for each component?

**Comment by prabhu** [comment]
@tech-surajrk What you're describing is a problem that is a bit more complex to solve. Essentially we need a mechanism to match an artifact to its identifier. There are a few techniques.

- Parse the META-INF/MANIFEST.MF in the jar file and try. Quite imprecise and doesn't work for legacy jars
- Have a hash of all jars in the world and compare to find all the properties. This approach is taken by [trivy](https://github.com/aquasecurity/trivy-java-db) and a few commercial tools. It might work, but not fool-proof
- Have a semantic fingerprint of all jars in the world and try to predict the closest match. Theoretical and coming to cdxgen this year if all goes well

So at this point in time, trivy is your best bet.

**Comment by Eiim** [comment]
I'd like to mention [syfts' approach](https://github.com/anchore/syft/tree/098c255a2dae0ce035d60dce0b7cb0324bf95063/syft/pkg/cataloger/java), which uses the filename as well as MANIFEST.MF. It's not perfect but it seems much more accurate than cdxgen, as most jars use the `name-version.jar` convention. I also strongly suspect that cdxgen's manifest parsing logic could be much improved, although vendors do lots of weird, out-of-spec things with manifest files so I don't necessarily blame cdxgen for having issue parsing these.

**Comment by prabhu** [comment]
@Eiim cdxgen also uses the [filename](https://github.com/CycloneDX/cdxgen/blob/master/utils.js#L4564) as a fallback. Could you file a ticket on cdxgen about the parsing logic improvements? Happy to take it up.
